pkg/provisioner/ironic/raid.go

```diff
@@ -251,20 +254,29 @@ func BuildRAIDCleanSteps(raidInterface string, target *metal3v1alpha1.RAIDConfig
 	return
 }
 
-func checkRAIDConfigure(raidInterface string, raid *metal3v1alpha1.RAIDConfig) error {
-	switch raidInterface {
-	case noRAIDInterface:
-		if raid != nil && (len(raid.HardwareRAIDVolumes) != 0 || len(raid.SoftwareRAIDVolumes) != 0) {
-			return fmt.Errorf("raid settings are defined, but the node's driver %s does not support RAID", raidInterface)
-		}
-	case softwareRAIDInterface:
-		if raid != nil && len(raid.HardwareRAIDVolumes) != 0 {
-			return fmt.Errorf("node's driver %s does not support hardware RAID", raidInterface)
-		}
-	default:
-		if raid != nil && len(raid.HardwareRAIDVolumes) == 0 && len(raid.SoftwareRAIDVolumes) != 0 {
-			return fmt.Errorf("node's driver %s does not support software RAID", raidInterface)
-		}
+func checkRAIDInterface(raidInterface string, target, actual *metal3v1alpha1.RAIDConfig) (string, error) {
+	if target == nil {
+		return raidInterface, nil
+	}
+
+	if raidInterface == noRAIDInterface && len(target.HardwareRAIDVolumes) != 0 {
+		return "", fmt.Errorf("RAID settings are defined, but the node's driver %s does not support RAID", raidInterface)
+	}
+
+	// This is not a real case since no BMC driver defaults to agent, but we add it for consistency
+	if raidInterface == softwareRAIDInterface && len(target.HardwareRAIDVolumes) != 0 {
+		return "", fmt.Errorf("hardware RAID settings are defined, but the node's driver %s only supports software RAID", raidInterface)
 	}
-	return nil
+
+	// If software RAID is requested, change the RAID interface.
+	if len(target.SoftwareRAIDVolumes) != 0 {
+		return softwareRAIDInterface, nil
+	}
+
+	// If software RAID is being deleted, also change the interface.
+	if actual != nil && len(target.SoftwareRAIDVolumes) == 0 && len(actual.SoftwareRAIDVolumes) != 0 {
```

Comment:
Second condition is always true here. Should it be HardwareRAIDVolumes?

Reply:
Mmm, no, how are hardware volumes related? This handles the case when we have some volumes created, and the user wants to delete them, thus setting target to [].

Reply:
As it stands if you're changing from Software RAID (in actual) to Hardware RAID (in target) then it'll return softwareRAIDInterface, which is surprising. (I'd find it easier to believe that there is _no_ right answer here than that this is the right answer.)

Reply:
> As it stands if you're changing from Software RAID (in actual) to Hardware RAID (in target) then it'll return softwareRAIDInterface, which is surprising.

There is no mention of hardware RAID here, I don't quite understand what you're referring to. This logic triggers when we have an existing software RAID (in actual), and the user requests no software RAID volumes (in target). To remove software RAID, we need to use the software RAID interface, otherwise it will stuck forever.

Reply:
Sorry, reading my comment again I think it was backwards. The issue is when you change from hardware RAID to software RAID.

Two configs are passed here, `target` which is the RAID setting we want to configure, and `actual`, which is the RAID setting that we already configured the host with.

If the host _was_ configured with hardware RAID and we're removing it, we use the hardware RAID driver to do so, as you'd expect. But if we're simultaneously replacing the hardware RAID config with a software RAID config, then we'll try to use the software RAID driver to remove the hardware RAID config, which is surprising.

Reply:
If this is intentional then you can delete the `len(target.SoftwareRAIDVolumes) == 0` part because it's already checked on line 273.

Reply:
I'll leave a FIXME and update the docs.